Improve derive syntax

pksunkara opened the pull request:
- [x] Sqlite tests
- [x] Postgres tests
- [x] Mysql tests
- [x] Comprehensive compile errors tests
- [x] Deprecated syntax support
- [x] Comprehensive compile warnings tests
- [x] `#[column_name = "QUERY PLAN"]` support
- [x] `TypePath` for `sql_type` and `foreign_derive`?
- [x] Remove `changeset_options` keyword

weiznich reviewed (changes requested):
That looks quite good :+1: Left a first bunch of comments. There are likely a few more things which need to change later on or I've missed now.

pksunkara commented:
Still some work to do. I just wanted to create a draft PR to show some progress. You can see that the commits are from 2 months ago :)

Regarding duplicated attributes, neither clap nor other repos which follow this overridable pattern has warnings. And I actually think it's good like that because it might open up some modularized use cases for end users.

pickfire commented on diesel_derives/src/attrs.rs:
```diff
@@ -0,0 +1,192 @@
+        .get_ident()
+        .map_or(None, |ident| match &*ident.to_string() {
+            "table_name" | "primary_key" | "column_name" | "sql_type" | "changeset_options" => {
+                emit_warning!(ident, "#[{}...] attribute form is deprecated", ident);
```
I think it could be a suggestion, not a note.

pksunkara replied:
@pickfire I didn't know rust compiler has suggestions. Can you please give me more info?

pickfire replied:
It is in rustc but not in proc_macro_error. Well, written by creepyskeleton maybe both of you all are not aware of it? https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=suggestions#suggestions

The diagnostic levels supported are also different https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=suggestions#diagnostic-levels

pksunkara replied:
Ah `span_suggestion` hasn't made into https://doc.rust-lang.org/proc_macro/struct.Diagnostic.html which is what `proc_macro_error` sets it's target as. cc @creepyskeleton

weiznich commented:
@pksunkara Are you interested in finishing this PR in the next few weeks?

pksunkara commented:
Yup, I was just rebasing this.

pksunkara commented:
@weiznich Why does selectable derive have `sql_type` attribute helper? [reference](https://github.com/diesel-rs/diesel/blob/master/diesel_derives/src/lib.rs#L771). I don't see it being used anywhere in the selectable expansion.

Also, since this is a new derive, we don't need backward compatibility for this and remove the old attribute helpers, right?

weiznich commented:
Yes, `sql_type`  is not used there.

> Also, since this is a new derive, we don't need backward compatibility for this and remove the old attribute helpers, right?

Sounds reasonable :+1:

pksunkara commented:
@weiznich Can you please give a preliminary look and see if you are okay with the format of the new attributes?

pksunkara commented:
CI is weird. I am locally on 1.48.0 and am unable to reproduce [this](https://github.com/diesel-rs/diesel/pull/2626/checks?check_run_id=3005203775#step:7:160) error at all. Or the `spanned::Spanned` error.

weiznich reviewed (commented):
I've done another round of review and left some comments on places where I have questions or small suggestions.

General points to discuss:
* As far as I see there is currently no way to emit a deprecation warning on a build not using rustc nightly. That's bad, we should find some way to "fix" that, even if that means to emit some "warnings" via `eprintln!`
* The `get_parent()` error for the minimal rust version build is caused by `-Z minimal-versions`, which just means that the minimal syn version that is specified do not have the corresponding method. This should be fixed by just increasing the minimal supported syn version to some version that has this method (at least syn 1.0.3 as far as I can see)
* There are a few places the old deprecated attributes are not converted. The CI mostly did emit warnings for them, they should be fixed
* I'm not sure about the `spanned::Spanned` error, are you really sure that this trait is really needed in that scope for all backends?

pksunkara commented on diesel/src/associations/mod.rs:
```diff
@@ -1,23 +1,23 @@
 //! #[derive(Identifiable, Queryable, Associations, PartialEq, Debug)]
-//! #[belongs_to(User)]
-//! #[table_name = "posts"]
+//! #[diesel(belongs_to(User))]
```
I don't mind either. I kept them separate because it was easy to do the refactoring by changing stuff in place using "find & replace"

weiznich replied:
Keeping them separate for refactoring is a good argument. In that way we don't clutter our git blame with unrelated changes. That means having them on different lines is preferred at least in diesel itself.

pksunkara commented on diesel/src/mysql/types/mod.rs:
```diff
@@ -198,5 +198,5 @@ impl HasSqlType<Unsigned<BigInt>> for Mysql {
 )]
 #[derive(Debug, Clone, Copy, Default, QueryId, SqlType)]
-#[mysql_type = "DateTime"]
+#[diesel(mysql_type("DateTime"))]
```
We needed `postgres_type()`. I wanted to keep the other types with similar looking syntax.

weiznich replied:
That's a good argument, but then I would go with something like this: `#[diesel(mysql_type(name = "DateTime"))]` as this looks much more consistent than.

pksunkara commented on diesel_derives/src/attrs.rs:
```diff
@@ -0,0 +1,282 @@
+    SqlType(Ident, Type),
+
+    // ident(ident = bool,*)
+    ChangesetOptions(Ident, Punctuated<(Ident, LitBool), Comma>),
```
Yeah, that check is being done in `model.rs`. The syntax allows adding more options in the future.

pksunkara commented on diesel_derives/src/attrs.rs:
```diff
@@ -0,0 +1,282 @@
+    // ident(type_path, ident = ident,*)
+    BelongsTo(Ident, TypePath, Punctuated<(Ident, Ident), Comma>),
+
+    // ident("lit_str")
+    MysqlType(Ident, LitStr),
+    SqliteType(Ident, LitStr),
```
Will be discussed and resolved above.

pksunkara commented on diesel_derives/src/lib.rs:
```diff
@@ -17,22 +17,20 @@
     clippy::items_after_statements,
     clippy::used_underscore_binding
 )]
-#![cfg_attr(feature = "nightly", feature(proc_macro_diagnostic, proc_macro_span))]
```
I am not really sure. There's a weird `cfg_attr(not(use_fallback))` which I don't understand.

weiznich replied:
Yes it is automatically enabled for nightly. They do a check for the current compiler version as part of their [build.rs](https://gitlab.com/CreepySkeleton/proc-macro-error/-/blob/55d609ba01a10a6ab05de51f21d71e622bb87c3f/build.rs#L2-4) file.

pksunkara commented on diesel_derives/tests/queryable_by_name.rs:
```diff
@@ -66,37 +69,37 @@ fn struct_with_no_table() {
-        data
-    );
-}
+// #[test]
+// fn struct_with_non_ident_column_name() {
+//     #[derive(Debug, Clone, PartialEq, Eq, QueryableByName)]
+//     struct QueryPlan {
+//         #[diesel(sql_type = diesel::sql_types::Text)]
+//         #[diesel(column_name = "QUERY PLAN")]
+//         qp: String,
+//     }
+
+//     let conn = &mut connection();
+//     let data = sql_query("SELECT 'some plan' AS \"QUERY PLAN\"").get_result(conn);
+//     assert_eq!(
+//         Ok(QueryPlan {
+//             qp: "some plan".to_string()
+//         }),
+//         data
+//     );
+// }
```
Yeah, I made `column_name` take `ident`. I saw this too late. I would need to allow `column_name` to take `lit_str` too.

weiznich replied:
This remains to be fixed/reenabled

pksunkara commented:
I haven't gone through the error/warning messages yet for the compile tests. I should have specifically said not to review those files inorder to save some time.

pksunkara commented on diesel_derives/src/lib.rs:
```diff
@@ -325,21 +325,21 @@ pub fn derive_identifiable(input: TokenStream) -> TokenStream {
-    );
-    expand_proc_macro(input, valid_grouping::derive)
-}
+// #[doc(hidden)]
+// #[proc_macro_derive(NonAggregate)]
+// pub fn derive_non_aggregate(input: TokenStream) -> TokenStream {
+//     eprintln!(
+//         "#[derive(NonAggregate)] is deprecated. Please use \
+//          `#[derive(ValidGrouping)]` instead.)"
+//     );
+//     expand_proc_macro(input, valid_grouping::derive)
+// }
```
This is added in v2 and deprecated at the same time. So, I want to remove it completely after confirming with you.

weiznich replied:
Then it's fine to remove it :+1:

pksunkara commented on diesel_derives/src/model.rs:
```diff
@@ -1,69 +1,179 @@
-            .into()
-        })
+    pub fn table_name(&self) -> &Path {
+        self.table_name.as_ref().unwrap_or_else(|| &self.name)
```
We want to know when the `table_name` attribute is actually specified or not.

weiznich commented:
> As far as I see there is currently no way to emit a deprecation warning on a build not using rustc nightly. That's bad, we should find some way to "fix" that, even if that means to emit some "warnings" via eprintln!

So the [Readme](https://gitlab.com/CreepySkeleton/proc-macro-error/-/tree/master) states:

> Warnings are emitted only on nightly, they are ignored on stable.

which is unfortunate, as we really want to show the deprecated message everywhere. I see a few options around this:

* Provide our own wrapping macro that emits warnings via `eprintln!` instead of just ignoring them. This likely requires a similar version check than done by `proc-macro-error`
* Try to fix that in `proc-macro-error` by maybe using `eprintln!` there?
* Emit an error instead? Maybe behind a feature flag? (Diesel already has a with/without-deprectated feature flag) (I'm not really sold on that option)

pksunkara commented:
Yeah, I am not exactly sure what we want to do regarding that. I am leaving it for the last thing to do though. Hopefully we will come up with something by then.

pksunkara commented:
@weiznich This is ready

weiznich commented on diesel_derives/src/insertable.rs:
```diff
@@ -85,23 +62,23 @@ pub fn derive(item: syn::DeriveInput) -> Result<proc_macro2::TokenStream, Diagno
-                    .error("`#[diesel(embed)]` cannot be combined with `#[diesel(serialize_as)]`"))
+            (Some(ty), true) => {
+                abort!(
+                    ty,
```
Shouldn't that point to the complete field or even to the derive itself instead of just the type?

pksunkara replied:
Agreed, I changed this to put to the span of the field. But I think there's a bug somewhere (either in syn or rust).

```
error: `#[diesel(embed)]` cannot be combined with `#[diesel(serialize_as)]`
  --> $DIR/embed_and_serialize_as_cannot_be_mixed.rs:22:5
   |
22 |     #[diesel(embed, serialize_as = SomeType)]
   |     ^
```

The span just points to the beginning of the derive and nothing else. Shall I leave it as that?

pksunkara replied:
Yeah, the above error message you see is what's happening when I use span of the field.

weiznich replied:
I will try to have a look at that this week. Otherwise thats something I would not consider to be a blocking issue for merging this PR.

pksunkara replied:
Okay, you can check this once I update the PR then.

weiznich replied:
I've pushed a fixed version. I've also fixed some copy paste error in a `#[derive(QueryableByName)]` error message.

weiznich commented on diesel_derives/src/parsers/postgres_type.rs:
```diff
@@ -0,0 +1,87 @@
+                schema, "expected `name` to be also present";
+                help = "make sure `name` is present, `#[diesel(postgres_type(name = \"...\", schema = \"{}\"))]`", schema.value()
+            );
+        } else if oid.is_some() && array_oid.is_some() {
```
Seems like clippy would prefer a `match` here instead

weiznich commented on diesel_derives/src/parsers/postgres_type.rs:
```diff
@@ -0,0 +1,87 @@
+        } else {
+            abort!(
+                input.span(),
+                "expected `oid` and `array_oid` attribute or `name` attribute"
```
Can we word this message in such a way that it shows concrete instanciations?
Something like
"note: `#[postgres(name = "whatever")]` or `#[postgres(oid=42, array_oid = 54)]` is expected?

weiznich replied:
```suggestion
                "expected `oid` and `array_oid` attribute or `name` attribute",
                help = "The correct format looks like either `#[diesel(name = "\…\", schema = "\…\")]` or `#[diesel(oid = …, array_oid = …)]`"
```

weiznich commented on diesel_derives/src/util.rs:
```diff
@@ -1,8 +1,28 @@
+use model::Model;
+
+pub fn unknown_attribute(name: &Ident) -> ! {
+    abort!(name, "unknown attribute")
```
Can we print the accepted values here?

weiznich commented on guide_drafts/trait_derives.md:
```diff
@@ -27,19 +27,12 @@ Diesel's code generation derives are for safely building SQL queries.
 [ORM]: https://en.wikipedia.org/wiki/Object-relational_mapping
 [CRUD]: https://en.wikipedia.org/wiki/Create,_read,_update_and_delete
 
-- [Queryable](#queryable)
-- [QueryableByName](#queryablebyname)
-- [Insertable](#insertable)
-- [Identifiable](#identifiable)
-- [AsChangeset](#aschangeset)
-- [Associations](#associations)
-
```
Any reason why those links are removed?

weiznich reviewed (commented):
Thanks for the update :+1:
This looks much better now. I left a few comments where I noticed small things. I think most of them should be easy to fix, some will likely require just a clarification.

pksunkara commented on diesel_compile_tests/rust-toolchain:
```diff
@@ -1 +0,0 @@
-nightly-2021-04-01
```
The `rust-toolchain` from parent dir is already defaulted to `1.51.0` and it works here.

weiznich commented:
I should probably comment here to say that I've a rebased version of this locally, which also fixes at least some of the remarks. I can try to clean it up a bit tomorrow and push it afterwards.

pksunkara commented:
> I should probably comment here to say that I've a rebased version of this locally, which also fixes at least some of the remarks. I can try to clean it up a bit tomorrow and push it afterwards.

Oh, I was just about to do the rebase for the final push this week to get this merged. I will wait for you. Would you mind if you create a new commit for your changes instead of bundling them up with my rebased commits?

weiznich commented:
I've pushed my rebased version. Hopefully this is a good starting point for fixing the remaining issues.

pksunkara commented on diesel/src/type_impls/primitives.rs:
```diff
@@ -55,7 +55,10 @@ mod foreign_impls {
     #[derive(FromSqlRow)]
-    #[cfg_attr(any(feature = "mysql_backend", feature = "postgres"), derive(AsExpression))]
+    #[cfg_attr(
+        any(feature = "mysql_backend", feature = "postgres"),
```
To check: Should `postgres` here be `postgres_backend`?

weiznich replied:
That should be `postgres_backend`, right.
```suggestion
        any(feature = "mysql_backend", feature = "postgres_backend"),
```

Generally speaking anything postgres specific that does not rely on `libpq` should be behind the `postgres_backend`, so anything that is not related to `PgConnection`.

pksunkara commented on diesel_compile_tests/tests/fail/derive_deprecated/deprecated_belongs_to.stderr:
```diff
@@ -0,0 +1,138 @@
+warning: #[belongs_to] attribute form is deprecated
```
Not really. I was just printing the warning as soon as I see it. And since I was testing bad formats on the old deprecated format too, you would see the warning first and then the error.

Your change fixed that because we now only warn if the format is okay. So, we don't see this warning anymore.

pksunkara commented on diesel_derives/src/as_expression.rs:
```diff
@@ -1,30 +1,30 @@
-        .into_iter()
-        .filter_map(|attr| attr.ty_value().map_err(Diagnostic::emit).ok());
+    if model.sql_types.is_empty() {
+        abort_call_site!(
```
I think it's better if we error out in this case because we don't people (who generally ignore warnings) to forget to put sql type for this derive (which is what this derive is all about in the first place).

weiznich replied:
:+1:

pksunkara commented:
I am not sold on the commit you pushed since we have the span already at [`dba9212` (#2626)](https://github.com/diesel-rs/diesel/pull/2626/commits/dba9212ffcec0ed3ea251b98b1f055e91997bbd1#diff-b58e3f3e8bc5ea8562c930edf63e06ae933fb77b8cf669fc93c63c52b99c101bR12).

I will be able to take a deeper look tomorrow and also address the last comment that's left.

weiznich commented:
> I am not sold on the commit you pushed since we have the span already at dba9212 (#2626).

I think an important point to note here is that there are a few different spans involved here:

* The span of the field as whole, as returned by `SynField.span()`. That's the span that seems to point only to the beginning of the field.
* The span of each element in the attribute. So in a `#[diesel(embed, serialize_as = "…")]` attribute are two of those spans. One pointing to `embed` and one pointing to `serialize_as = "…"`. Those are stored in `ident_span`
* The span of a field attribute as whole thing. For  `#[diesel(embed, serialize_as = "…")]` this will point to `diesel(embed, serialize_as = "…")`. It seems like trying to point to the whole attribute just creates a single char span at the beginning 😞 . That needs to be a separate entry as there may be more than one attribute here per field. That stored in `attribute_span`
* The span of the field itself. I opted to use either the span of the field name (for structs with named fields) or the span of the field type (for tuple structs) for this. That stored in `field.span`

pksunkara commented:
I am going to separate the `AttributeSpanWrapper` work into a separate PR that will get opened up once this is merged as a fast follow. I think we need a little bit discussion on it. And I might have a better idea on how to do that.

I pulled the commit locally and saved it.

pksunkara commented on diesel_compile_tests/tests/fail/derive/bad_aggregate.stderr:
```diff
@@ -0,0 +1,5 @@
+error: expected `,`
```
It seems to be a general issue of syn especially with `Punctuated` type. It doesn't have the full context and thus only prints out parsing errors talking about the next token.

And this seems to be an issue with of all derives in the ecosystem. So, I am okay with leaving this as it is and expect users to recognise the error for what it is.

weiznich replied:
Thanks for checking this. In this case it seems like the best solution to ignore this for now and wait for someone that fixes the ecosystem.

weiznich commented on diesel_compile_tests/tests/fail/derive/bad_not_sized.rs:
```diff
@@ -0,0 +1,9 @@
+use diesel::expression::AsExpression;
+
+#[derive(AsExpression)]
+#[diesel(sql_type = bool)]
```
Minor nit:
```suggestion
```

(Mind that just applying this via Github's UI will not work, as we need to update the output as well)

weiznich commented on diesel_compile_tests/tests/fail/derive/bad_postgres_type.stderr:
```diff
@@ -0,0 +1,80 @@
+error: unexpected `oid` when `name` is present
+  --> $DIR/bad_postgres_type.rs:30:44
+   |
+30 | #[diesel(postgres_type(name = "foo", oid = 2, array_oid = 3))]
```
As a thing for the followup PR: This spans could also be improved using the same strategy as demonstrated for `deserialize_as`/`embed`

pksunkara commented:
Addressed all the comments now. The only question left from my original goal of the work is:

> `TypePath` for `sql_type`, `serialize_as`, `deserialize_as`?

Can you please give a look at https://docs.rs/syn/latest/syn/enum.Type.html and tell me which of those variants do we allow for the above mentioned attributes?

weiznich commented on diesel_derives/src/parsers/postgres_type.rs:
```diff
@@ -0,0 +1,90 @@
+    ) -> Result<Self> {
+        if let Some(name) = name {
+            if oid.is_some() {
+                abort!(oid, "unexpected `oid` when `name` is present");
```
```suggestion
                abort!(
                    oid, "unexpected `oid` when `name` is present",
                    help = "The correct format looks like either `#[diesel(name = "\…\", schema = "\…\")]` or `#[diesel(oid = …, array_oid = …)]`"
                );
```

weiznich commented on diesel_derives/src/parsers/postgres_type.rs:
```diff
@@ -0,0 +1,90 @@
+            if oid.is_some() {
+                abort!(oid, "unexpected `oid` when `name` is present");
+            } else if array_oid.is_some() {
+                abort!(array_oid, "unexpected `array_oid` when `name` is present");
```
```suggestion
                abort!(
                    array_oid, "unexpected `array_oid` when `name` is present",
                    help = "The correct format looks like either `#[diesel(name = "\…\", schema = "\…\")]` or `#[diesel(oid = …, array_oid = …)]`"
                 );
```

weiznich commented on diesel_derives/src/parsers/mysql_type.rs:
```diff
@@ -0,0 +1,45 @@
+        if let Some(name) = name {
+            Ok(MysqlType { name })
+        } else {
+            abort!(input.span(), "expected attribute `name`");
```
```suggestion
            abort!(
                input.span(), "expected attribute `name`",
                help = "The correct format looks like `#[diesel(mysql_type(name = \"abc\"))]`"
            );
```

weiznich commented on diesel_derives/src/parsers/sqlite_type.rs:
```diff
@@ -0,0 +1,45 @@
+        if let Some(name) = name {
+            Ok(SqliteType { name })
+        } else {
+            abort!(input.span(), "expected attribute `name`");
```
```suggestion
            abort!(
                input.span(), "expected attribute `name`",
                help = "The correct format looks like `#[diesel(sqlite_type(name = \"abc\"))]`"
            );
```

weiznich reviewed (commented):
Just a bunch of minor notes. Those should be easy to fix. Otherwise I think this now looks really good :tada:.

One thing is still missing: A changelog entry + probably some draft section for the upgrade guide (Just some few words that we updated the attribute format…)

weiznich commented:
> Can you please give a look at https://docs.rs/syn/latest/syn/enum.Type.html and tell me which of those variants do we allow for the above mentioned attributes?

I would just accept [`TypePath`](https://docs.rs/syn/latest/syn/struct.TypePath.html) there and nothing else. If that turns out to be not enough we can easily include more variants later on.

pksunkara commented:
Address review and used `TypePath`

weiznich reviewed (approved):
Let's merge this :tada:

weiznich commented:
@pksunkara I want to use this opportunity to thank your for your long term effort here. I really appreciate this. :heart:

pksunkara commented:
No problem. I just wish I had the time to actually focus on this properly in a straight forward manner instead of doing over a long time.
